
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../Utilities/">
      
      
        <link rel="next" href="../tool_params/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.20">
    
    
      
        <title>Configuration - mikrokondo</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.e53b48f4.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#configuration" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="mikrokondo" class="md-header__button md-logo" aria-label="mikrokondo" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            mikrokondo
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Configuration
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../.." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../installation/" class="md-tabs__link">
          
  
  
  Usage

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../troubleshooting/FAQ/" class="md-tabs__link">
          
  
  
  Troubleshooting

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../workflows/CleanAssemble/" class="md-tabs__link">
          
  
  
  Workflows

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../subworkflows/assemble_reads/" class="md-tabs__link">
          
  
  
  Subworkflows

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="mikrokondo" class="md-nav__button md-logo" aria-label="mikrokondo" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    mikrokondo
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Usage
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Usage
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Running MikroKondo
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../examples/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Command Line Examples
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Utilities/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Utilities
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Configuration
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Configuration
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#configuration-files-overview" class="md-nav__link">
    <span class="md-ellipsis">
      Configuration files overview
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Configuration files overview">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#base-configuration-confbaseconfig" class="md-nav__link">
    <span class="md-ellipsis">
      Base configuration (conf/base.config)
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#hardcoded-tool-configuration-nextflowconfig" class="md-nav__link">
    <span class="md-ellipsis">
      Hardcoded tool configuration (nextflow.config)
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#quality-control-report-configuration" class="md-nav__link">
    <span class="md-ellipsis">
      Quality control report configuration
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Quality control report configuration">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#qcreport-field-description" class="md-nav__link">
    <span class="md-ellipsis">
      QCReport field description
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#example-adding-quality-control-data-for-salmonella" class="md-nav__link">
    <span class="md-ellipsis">
      Example adding quality control data for Salmonella
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#quality-control-fields" class="md-nav__link">
    <span class="md-ellipsis">
      Quality Control Fields
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#locidex-manifest-file" class="md-nav__link">
    <span class="md-ellipsis">
      Locidex Manifest File
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Locidex Manifest File">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#how-automated-selection-works" class="md-nav__link">
    <span class="md-ellipsis">
      How automated selection works.
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tool_params/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Tool Specific Parameters
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Troubleshooting
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Troubleshooting
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../troubleshooting/FAQ/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    FAQ
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Workflows
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Workflows
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../workflows/CleanAssemble/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Clean Assemble
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../workflows/PostAssembly/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Post assembly
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Subworkflows
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Subworkflows
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../subworkflows/assemble_reads/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Assembly
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../subworkflows/bin_contigs/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Bin Contigs
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../subworkflows/clean_reads/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Read Quality Control
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../subworkflows/determine_species/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Determine Species
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../subworkflows/genomes_annotate/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Genome Annotation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../subworkflows/hybrid_assembly/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Hybrid Assembly
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../subworkflows/input_check/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Input Verification
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../subworkflows/polish_assemblies/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Assembly Polishing
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../subworkflows/qc_assembly/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Assembly Quality Control
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../subworkflows/subtype_genome/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Genome Sub-typing
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  



<h1 id="configuration">Configuration</h1>
<h2 id="configuration-files-overview">Configuration files overview</h2>
<p>The following files contain configuration settings:</p>
<ul>
<li>
<p><code>conf/base.config</code>: Where cpu, memory and time parameters can be set for the different workflow processes. <strong>You will likely need to adjust parameters within this file for your computing environment</strong>.</p>
</li>
<li>
<p><code>conf/modules.config</code>: contains error strategy, output director structure and execution instruction parameters. <strong>It is unadvised to alter this file unless involved in pipeline development, or tuning to a system.</strong></p>
</li>
<li>
<p><code>conf/equivalent_taxa.json</code>: Contains a set of keys to arrays containing the query notes from the mash sketch denoting "equivalent taxa". This typically contains a list of organisms that are genetically similar but phenotypically distinct, as mobile elements or genome segments may be shared across organisms. e.g. <em>Shigella</em> and <em>Escherichia</em></p>
</li>
<li>
<p><code>nextflow.config</code>: contains default tool settings that tie to CLI options. These options can be directly set within the <code>params</code> section of this file in cases when a user has optimized their pipeline usage and has identified the flags they will use every time the pipeline is run.</p>
</li>
</ul>
<h3 id="base-configuration-confbaseconfig">Base configuration (conf/base.config)</h3>
<p>Within this file computing resources can be configured for each process. Mikrokondo uses labels to define resource requirements for each process, here are their definitions:</p>
<ul>
<li><code>process_single</code>: processes requiring only a single core and low memory (e.g., listing of directories).</li>
<li><code>process_low</code>: processes that would typically run easily on a small laptop (e.g., staging of data in a Python script).</li>
<li><code>process_medium</code>: processes that would typically run on a desktop computer equipped for playing newer video games (Memory or computationally intensive applications that can be parallelized, e.g., rendering, processing large files in memory or running BLAST).</li>
<li><code>process_high</code>: processes that would typically run on a high performance desktop computer (Memory or computationally intensive application, e.g., performing <em>de novo</em> assembly or performing BLAST searches on large databases).</li>
<li><code>process_long</code>: modifies/overwrites the amount of time allowed for any of the above processes to allow for certain jobs to take longer (e.g., performing <em>de novo</em> assembly with less computational resources or performing global alignments on divergent sequences).</li>
<li><code>process_high_memory</code>: modifies/overwrites the amount of memory given to any process and grant significantly more memory to any process (Aids in metagenomic assembly or clustering of large datasets).</li>
</ul>
<p>For actual resource amounts allotted to each process definition, see the <code>conf/base.config</code> file <em>Process-specific resource requirements</em> section.</p>
<h3 id="hardcoded-tool-configuration-nextflowconfig">Hardcoded tool configuration (nextflow.config)</h3>
<p>All Command line arguments and defaults can be set and/or altered in the <code>nextflow.config</code> file, <em>params</em> section. For a full list of parameters to be altered please refer to the <code>nextflow.config</code> file in the repo. Some common arguments have been listed in the <a href="/usage/useage/#common-command-line-arguments">Common command line arguments</a> section of the docs and further description of tool parameters can also be found in <a href="/usage/tool_params/">tool specific parameters</a>.</p>
<blockquote>
<p><strong>Example:</strong> if your laboratory typically sequences using Nanopore chemistry "r1041_e82_400bps_hac_v4.2.0", the following code would be substituted in the <em>params</em> section of the <code>nextflow.config</code> file:</p>
<p><code>nanopore_chemistry = "r1041_e82_400bps_hac_v4.2.0" // Note the quotes around the value</code></p>
<p>With this change, you would no longer need to explicitly state the nanopore chemistry as an extra CLI argument when running mikrokondo.</p>
</blockquote>
<h2 id="quality-control-report-configuration">Quality control report configuration</h2>
<blockquote>
<p><strong>WARNING:</strong> Tread carefully here, as this will require modification of the <code>nextflow.config</code> file. <strong>Make sure you have saved a back up of your <code>nextflow.config</code> file before playing with these option</strong></p>
</blockquote>
<h3 id="qcreport-field-description">QCReport field description</h3>
<p>The section of interest is the <code>QCReport</code> fields in the params section of the <code>nextflow.config</code>. There are multiple sections with values that can be modified or you can add data for a different organism. The default values in the pipeline are set up for <strong>Illumina data</strong> so you may need to adjust settingS for Nanopore or Pacbio data.</p>
<blockquote>
<p><strong>WARNING:</strong> It is best to only add <code>QCReport</code> values to mikrokondo, not remove existing ones as this may raise errors in the subtyping module.</p>
</blockquote>
<p>An example of the QCReport structure is shown below. With annotation describing the values.</p>
<blockquote>
<p><strong>NOTE:</strong> The values below do not affect the running of the pipeline, these values only affect the final quality messages output by the pipeline.</p>
<p><strong>NOTE:</strong> The term JSON path is used below to indicate the ordered set of keys required to reach a particular value in an aggregated JSON file created by mikrokondo. For example if you wished to specify the JSON path to the ECTyper species identification in mikrokondo you would enter <code>["ECTyperSubtyping", "0", "Species"]</code> as the JSON path which would point to a samples ECTyper speciation field in the following JSON structure from the <code>final_report.json</code> file generated by mikrokondo:</p>
</blockquote>
<pre><code>  &quot;ECTyperSubtyping&quot;: {
    &quot;0&quot;: {
      &quot;Species&quot;: &quot;Escherichia coli&quot;
    }
  }
</code></pre>
<p>If you need to find a the JSON path required to point to a value, you can either look at the sample JSON output in <code>FinalReports/Aggregated/Json/final_report.json</code> file generated by mikrokondo to determine the required sequence of keys. However it is simpler to find the required JSON path value by looking in the <code>FinalReports/Sample/Json/final_report_flattened.json</code> file where you will see the key path listed and delimited by '.' e.g. <code>ECTyperSubtyping.0.Species</code> which would be re-written as <code>["ECTyperSubtyping, "0", "Species"]</code>.</p>
<pre><code>QCReport {
    escherichia // Generic top level name fo the field, it is name is technically arbitrary but it nice field name keeps things organized
    {
        search = &quot;Escherichia&quot; // The phrase that is searched for in the species_top_hit field mentioned above. The search is for containment so if you wanted to look for E.coli and E.albertii you could just set the value to &quot;Escherichia coli&quot; or &quot;Escherichia albertii&quot;
        raw_average_quality = 30 // Minimum raw average quality of all bases in the sequencing data. This value is generated before the decontamination procedure.
        min_n50 = 95000 // The minimum n50 value allowed from quast
        max_n50 = 6000000 // The maximum n50 value allowed from quast
        min_nr_contigs = 1 // the minimum number of contigs a sample is allowed to have, a value of 1 works as a sanity check
        max_nr_contigs = 500 // The maximum number of contigs the organism in the search field is allowed to have. to many contigs could indicate a bad assembly or contamination
        min_length = 4500000 // The minimum genome length allowed for the organism specified in the search field
        max_length = 6000000 // The maximum genome length the organism in the search field is allowed to have
        max_checkm_contamination = 3.0 // The maximum level of allowed contamination allowed by CheckM
        min_average_coverage = 30 // The minimum average coverage allowed
        // If you wish to make use of IDField and IDTool you will need to set both values
        IDField = null // null|JSON path to relevant file results if null the mash or kraken2 results will be used
        IDTool = null // null|tool name used to create file result if null mash or kraken2 will be written
        // If PrimaryTypeID is set PrimaryTypeIDMethod must be as well
        PrimaryTypeID: null|JSON Path, Optional primary type to have displayed e.g. serotype
        PrimaryTypeIDMethod: null|String, Method used for Primary type e.g. ECTyper
        // If SecondaryTypeID is set SecondaryTypeIDMethod must be as well
        SecondaryTypeID: null|JSON Path, Secondary type information e.g. Clonal Complex
        SecondaryTypeIDMethod: null|String, method used for secondary type information e.g. 7 Gene
    }
    // DO NOT REMOVE THE FALLTRHOUGH FIELD AS IT IS NEEDED TO CAPTURE OTHER ORGANISMS
    fallthrough // The fallthrough field exist as a default value to capture organisms where no quality control data has been specified
    {
        search = &quot;No organism specific QC data available.&quot;
        raw_average_quality = 30
        min_n50 = null
        max_n50 = null
        min_nr_contigs = null
        max_nr_contigs = null
        min_length = null
        max_length = null
        max_checkm_contamination = 3.0
        min_average_coverage = 30
        IDField = null
        IDTool = null
        PrimaryTypeID = null
        PrimaryTypeIDMethod = null
        SecondaryTypeID = null
        SecondaryTypeIDMethod = null
    }
}
</code></pre>
<h3 id="example-adding-quality-control-data-for-salmonella">Example adding quality control data for <em>Salmonella</em></h3>
<p>If you wanted to add quality control data for <em>Salmonella</em> you can start off by using the template below:</p>
<pre><code>VAR_NAME { // Replace VAR name with the genus name of your sample, only use ASCII (a-zA-Z) alphabet characters in the name and replace spaces, punctuation and other special characters with underscores (_)
    search = &quot;Search phrase&quot; // Search phrase for your species top_hit, Note the quotes
    raw_average_quality = // 30 is a default value please change it as needed
    min_n50 = // Set your minimum n50 value
    max_n50 = // Set a maximum n50 value
    min_nr_contigs = // Set a minimum number of contigs
    max_nr_contigs = // The maximum number of contigs
    min_length = // Set a minimum genome length
    max_length = // set a maximum genome length
    max_checkm_contamination = // Set a maximum level of contamination to use
    min_average_coverage = // Set the minimum coverage value
    IDField = null // Set a Json path found in the flattened report generated by mikrokondo to your tool result
    IDTool = null // Set a custom name for the ID method you want to set
}
</code></pre>
<p>For <em>Salmonella</em> I would fill in the values like so.</p>
<pre><code>salmonella {
    search = &quot;Salmonella&quot;
    raw_average_quality = 30
    min_n50 = 95000
    max_n50 = 6000000
    min_nr_contigs = 1
    max_nr_contigs = 200
    min_length = 4400000
    max_length = 6000000
    max_checkm_contamination = 3.0
    min_average_coverage = 30
    IDField = [params.sistr.report_tag, &quot;0&quot;, &quot;Serovar&quot;]
    IDTool = &quot;SISTR&quot;
}
</code></pre>
<p>After having my values filled out, I can simply add them to the QCReport section in the <code>nextflow.config</code> file.</p>
<pre><code>    QCReport {
        escherichia {
            search = &quot;Escherichia coli&quot;
            raw_average_quality = 30
            min_n50 = 95000
            max_n50 = 6000000
            min_nr_contigs = 1
            max_nr_contigs = 500
            min_length = 4500000
            max_length = 6000000
            max_checkm_contamination = 3.0
            min_average_coverage = 30
        } salmonella { // NOTE watch the opening and closing brackets
            search = &quot;Salmonella&quot;
            raw_average_quality = 30
            min_n50 = 95000
            max_n50 = 6000000
            min_nr_contigs = 1
            max_nr_contigs = 200
            min_length = 4400000
            max_length = 6000000
            max_checkm_contamination = 3.0
            min_average_coverage = 30
        }
        fallthrough {
            search = &quot;No organism specific QC data available.&quot;
            raw_average_quality = 30
            min_n50 = null
            max_n50 = null
            min_nr_contigs = null
            max_nr_contigs = null
            min_length = null
            max_length = null
            max_checkm_contamination = 3.0
            min_average_coverage = 30
        }
    }
</code></pre>
<h2 id="quality-control-fields">Quality Control Fields</h2>
<p>This section affects the behavior of the final summary quality control messages and is noted in the <code>QCReportFields</code> within the <code>nextflow.config</code>. <strong>I would advise against manipulating this section unless you really know what you are doing</strong>.</p>
<p>Each value in the QC report fields contains the following fields.</p>
<ul>
<li>Field name</li>
<li>path: path to the information in the summary report JSON</li>
<li>coerce_type: Type to be coerced too, can be a Float, Integer, or Bool</li>
<li>compare_fields: A list of fields corresponding to fields in the <code>QCReport</code> section of the <code>nextflow.config</code>. If two values are specified it will be assumed you wish to check that a value is in between a range of values.</li>
<li>comp_type: The comparison type specified, 'ge' for greater or equal, 'le' for less than or equal, 'bool' for true or false or 'range' for checking if a value is between two values.</li>
<li>on: A boolean value for disabling a comparison</li>
<li>low_msg: A message for if a value is less than its compared value (optional)</li>
<li>high_msg: A message for if value is above a certain value (optional)</li>
</ul>
<p>An example of what these fields look like is:</p>
<pre><code>QCReportFields {
    raw_average_quality {
        path = [params.raw_reads.report_tag, &quot;combined&quot;, &quot;qual_mean&quot;]
        coerce_type = 'Float'
        compare_fields = ['raw_average_quality']
        comp_type = &quot;ge&quot;
        on = true
        low_msg = &quot;Base quality is poor, resequencing is recommended.&quot;
    }
}
</code></pre>
<h2 id="locidex-manifest-file">Locidex Manifest File</h2>
<p>Automated selection allele calling databases is supported within mikrokondo. This is accomplished with the help of Locidex itself, which offers a utility to generate a <code>manifest.json</code> file.</p>
<p>The directory of a database set for Locidex contains the following structure as the <code>manifest.json</code> keeps track of the paths relative too the location of the manifest file itself:</p>
<pre><code>--|
  |- Database 1
  |- Database 2
  |- Database n
  |- manifest.json
</code></pre>
<p>An example <code>manifest.json</code> file can be found in the mikrokondo <a href="https://github.com/phac-nml/mikrokondo/tree/main/tests/data/databases/locidex_dbs">test data sets here</a>.</p>
<p>Internally the <code>manifest.json</code> contains the following structure. Modifications to what <code>locidex manifest</code> outputs can be made as long as all fields populated. In the below example the <code>manifest.json</code> file generated by locidex has been modified to create two separate entries for <em>Escherichia coli</em> and <em>Shigella</em>.</p>
<pre><code>{
  &quot;Salmonella&quot;: [
    {
      &quot;path&quot;: &quot;wgmlst_salmonella&quot;,
      &quot;config&quot;: {
        &quot;db_name&quot;: &quot;Salmonella&quot;,
        &quot;db_version&quot;: &quot;1.0.0&quot;,
        &quot;db_date&quot;: &quot;2024-03-17&quot;,
        &quot;db_author&quot;: &quot;Tester&quot;,
        &quot;db_desc&quot;: &quot;Salmonella Database&quot;,
        &quot;db_num_seqs&quot;: 51251,
        &quot;is_nucl&quot;: true,
        &quot;is_prot&quot;: true,
        &quot;nucleotide_db_name&quot;: &quot;nucleotide&quot;,
        &quot;protein_db_name&quot;: &quot;protein&quot;
      }
    }
  ],
  &quot;Escherichia coli&quot;: [
    {
      &quot;path&quot;: &quot;wgmlst_escherichia_shigella&quot;,
      &quot;config&quot;: {
        &quot;db_name&quot;: &quot;EC and Shigella&quot;,
        &quot;db_version&quot;: &quot;1.0.0&quot;,
        &quot;db_date&quot;: &quot;2024-04-30&quot;,
        &quot;db_author&quot;: &quot;Tester&quot;,
        &quot;db_desc&quot;: &quot;Shigella and E.coli&quot;,
        &quot;db_num_seqs&quot;: 57692,
        &quot;is_nucl&quot;: true,
        &quot;is_prot&quot;: true,
        &quot;nucleotide_db_name&quot;: &quot;nucleotide&quot;,
        &quot;protein_db_name&quot;: &quot;protein&quot;
      }
    }
  ],
  &quot;Shigella&quot;: [
    {
      &quot;path&quot;: &quot;wgmlst_escherichia_shigella&quot;,
      &quot;config&quot;: {
        &quot;db_name&quot;: &quot;EC and Shigella&quot;,
        &quot;db_version&quot;: &quot;1.0.0&quot;,
        &quot;db_date&quot;: &quot;2024-04-30&quot;,
        &quot;db_author&quot;: &quot;Tester&quot;,
        &quot;db_desc&quot;: &quot;Shigella and E.coli&quot;,
        &quot;db_num_seqs&quot;: 57692,
        &quot;is_nucl&quot;: true,
        &quot;is_prot&quot;: true,
        &quot;nucleotide_db_name&quot;: &quot;nucleotide&quot;,
        &quot;protein_db_name&quot;: &quot;protein&quot;
      }
    }
  ],
  &quot;Listeria Monocytogenes&quot;: [
    {
      &quot;path&quot;: &quot;wgmlst_listeria&quot;,
      &quot;config&quot;: {
        &quot;db_name&quot;: &quot;Listeria Monocytogenes wgMLST&quot;,
        &quot;db_version&quot;: &quot;1.0.0&quot;,
        &quot;db_date&quot;: &quot;2024-04-16&quot;,
        &quot;db_author&quot;: &quot;Tester&quot;,
        &quot;db_desc&quot;: &quot;Listeria Monocytogenes wgMLST&quot;,
        &quot;db_num_seqs&quot;: 22404,
        &quot;is_nucl&quot;: true,
        &quot;is_prot&quot;: true,
        &quot;nucleotide_db_name&quot;: &quot;nucleotide&quot;,
        &quot;protein_db_name&quot;: &quot;protein&quot;
      }
    }
  ]
}
</code></pre>
<h3 id="how-automated-selection-works">How automated selection works.</h3>
<p>Mikrokondo is able to identify the species that a sample represents internally, but in order to identify the correct WgMLST scheme to use for allele calling the top-level key in the <code>manifest.json</code> file must be a name that can be parsed from the speciation output of Mash or Kraken2 e.g. <em>Salmonella enterica</em>, <em>Campylobacter_A anatolicus</em>, <em>Escherichia</em> etc.</p>
<blockquote>
<p><strong>Note:</strong> The database and organism names are not case sensitive.</p>
</blockquote>
<p>Mikrokondo will then be able to match the bacterial name outputs to what is in the <code>manifest.json</code>. In the following example below the three bacteria (<em>Salmonella enterica</em>, <em>Campylobacter_A anatolicus</em>, <em>Escherichia coli</em>) would all be matched to the correct scheme:</p>
<pre><code>{
  &quot;Salmonella&quot;: [
    ...
  ],
  &quot;Escherichia coli&quot;: [
    ...
  ],
  &quot;Campylobacter&quot;: [
    ...
  ]
}
</code></pre>
<p>This is because mikrokondo looks for the best exact match from the database names in the output species name. So spurious tokens like the <code>_A</code> in <em>Campylobacter_A anatolicus</em> would be removed and the <code>Campylobacter</code> database would be selected. For Salmonella, as the key value <code>Salmonella</code> overlaps entirely with the <em>Salmonella</em> of <em>Salmonella enterica</em> the Salmonella database would be selected. If There was a <code>Salmonella Enterica</code> database that would be selected over the generic <code>Salmonella</code> scheme. The <code>Escherichia coli</code> database would be selected for <em>Escherichia coli</em> as there they are a 100% match.</p>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tabs", "navigation.tabs.sticky", "navigation.sections", "navigation.expand", "navigation.top", "toc.integrate", "search.suggest", "search.highlight", "content.tabs.link", "content.code.annotation", "content.code.copy"], "search": "../../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.f55a23d4.min.js"></script>
      
    
  </body>
</html>